mkosi/__init__.py

```diff
@@ -1521,103 +1531,79 @@ def mount_overlay(
                f'upperdir={realroot}',
                f'workdir={workdir.name}']
 
-    do_mount("overlay", realroot, options, type="overlay", read_only=read_only)
-    mount_bind(realroot, root)
-    return realroot, workdir
-
-
-@complete_step("Cleaning up overlayfs")
-def clean_up_overlay(root: Path, realroot: Path, workdir: TempDir) -> None:
-    """Destroy the overlayfs structure set up by `mount_overlay`.
-
-    When this function returns, the contents of the root file system have been
-    moved into root, and `realroot` and `workdir` are gone.
-
-    If `realroot` is set, it means we mounted `root` twice: the first mount is
-    the overlayfs mount, and the second is a bind-mount to adjust the location
-    one level up. Thus we need unmount twice too; after the first unmount here,
-    the image remains mounted at `root`.
-    """
-
-    umount(root)
-    umount(realroot)
-
-    workdir.cleanup()
-
-    # Let's now move the contents of realroot into root
-    for entry in os.scandir(realroot):
-        os.rename(realroot / entry.name, root / entry.name)
-    realroot.rmdir()
+    try:
+        with workdir, mount("overlay", realroot, options=options, type="overlay", read_only=read_only), mount_bind(realroot, root):
+            yield root
+    finally:
+        with complete_step("Cleaning up overlayfs"):
+            # Let's now move the contents of realroot into root
+            for entry in os.scandir(realroot):
+                os.rename(realroot / entry.name, root / entry.name)
+            realroot.rmdir()
 
-    delete_whiteout_files(root)
+            delete_whiteout_files(root)
 
 
 @contextlib.contextmanager
 def mount_image(
     args: MkosiArgs,
     root: Path,
+    do_run_build_script: bool,
+    cached: bool,
     base_image: Optional[Path],  # the path to the mounted base image root
     loopdev: Optional[Path],
     image: LuksSetupOutput,
     root_read_only: bool = False,
 ) -> Iterator[None]:
-    with complete_step("Mounting image…"):
-
-        realroot: Optional[Path] = None
-        workdir: Optional[TempDir] = None
+    with complete_step("Mounting image…", "Unmounting image…"), contextlib.ExitStack() as stack:
```

Comment:
Here and below too…

Reply:
Adding line breaks to a `with` statements means using `\` at the end though? We can't yet use parentheses for multi line with because that got added after python 3.7 which is our minimum version atm